tests/python/pants_test/engine/test_objects.py

```diff
@@ -0,0 +1,13 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from pants.engine.objects import Collection
+from pants_test.test_base import TestBase
+
+
+class CollectionTest(TestBase):
+  def test_collection_iteration(self):
+    self.assertEqual([1, 2], [x for x in Collection.of(int)([1, 2])])
```

Comment:
It'd be nice to have a test that checks the element_type parts of Collection. It looks like Collection doesn't assert element types itself, which is fine, but it might be good to exercise that here too.

Reply:
I agree, we may want to pull out the backend and check types of Collection members differently at some point (or not), and it would be good to check that behavior since it is widely used already in the v2 engine. I will add this.

Reply:
I have added testing for this. The testing checks the error message as well, which is somewhat doing the same testing as in `util/test_objects.py`. I prefer to have this so that we can modify the error message as needed in the future without tying it just to the type checking error messages, as `Collection` is intended to be widely and easily used.